### PR TITLE
NamedTuple - Raises an error on non-iterable elements

### LINKED ISSUE #4369: NamedTuple.__init__ hangs with second argument as literal
Tested on '0.45.0' and '0.46.0dev0+5.gfddbdc5'

The following piece of code hangs forever when calling `NamedTupe.__init__` when second argument is a literal.

This should probably raise an error?

```python
from numba import types


types.NamedTuple(types.uint32, 2)
```

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Fix #4369

I still need to write some tests.

### CODE REVIEW

**Comment by esc** [comment]
@guilhermeleobas no problem, do let us know when you want us to review it.

**Comment by guilhermeleobas** [comment]
Hi @esc, I think this PR is ready for a review. 

CI is failing because the *Linux py35_np111_cov* was canceled.

**Comment by esc** [comment]
@guilhermeleobas thanks, I have scheduled it for a review.

File: numba/types/containers.py [hunk 1]
```diff
@@ -1,5 +1,7 @@
 from __future__ import print_function, division, absolute_import
 
+from collections import Iterable
+
 from .abstract import *
```
Comment (esc) [inline numba/types/containers.py]: not part of this PR, but I smell star imports that ought to be removed.

File: numba/tests/test_tuples.py [hunk 2]
```diff
@@ -361,6 +361,18 @@ def test_le(self):
...
 class TestNamedTuple(TestCase, MemoryLeakMixin):
 
+    def test_exception(self):
+        pyfunc = lambda: types.NamedTuple(types.uint32, 'p')
+        self.disable_leak_check()
+
+        with self.assertRaises(errors.TypingError) as raises:
+            pyfunc()
+
+        self.assertIn(
```
Comment (esc) [inline numba/tests/test_tuples.py]: The `assertIn` check should probably be indented.

File: numba/tests/test_tuples.py [hunk 3]
```diff
@@ -361,6 +361,18 @@ def test_le(self):
 
 class TestNamedTuple(TestCase, MemoryLeakMixin):
 
+    def test_exception(self):
+        pyfunc = lambda: types.NamedTuple(types.uint32, 'p')
```
Comment (esc) [inline numba/tests/test_tuples.py]: I am not sure we need the `pyfunc` here. The term `pyfunc` usually refers to the python original of an JIT compiled function, your use of this variable here is somewhat unusual and I think you can just call `types.NamedTuple`. in the `with` statement below.

File: numba/tests/test_tuples.py [hunk 4]
```diff
@@ -361,6 +361,18 @@ def test_le(self):
 
 class TestNamedTuple(TestCase, MemoryLeakMixin):
 
+    def test_exception(self):
```
Comment (esc) [inline numba/tests/test_tuples.py]: This might be named differently, like `test_exception_during_instantiation_with_non_iterable_types`, but maybe not as long. Maybe one of the  other core-devs has a better suggestion here.

**Comment by esc** [comment]
@guilhermeleobas thanks for taking care of this, it looks good to me know. I'll hand off to @stuartarchibald and @sklam for a final review and approval! Thanks!

File: numba/types/containers.py [hunk 5]
```diff
@@ -308,6 +310,10 @@ def key(self):
 class NamedTuple(_HeterogeneousTuple, BaseNamedTuple):
 
     def __init__(self, types, cls):
+
+        if not isinstance(types, Iterable):
+            raise TypingError("{} type is not iterable".format(type(types)))
```
Comment (stuartarchibald) [inline numba/types/containers.py]: I think this check would be better placed in the `_HeterogeneousTuple` class as a method and then it be called by the `__init__` of both `Tuple` and `NamedTuple` as they both require an `Iterable` for the `types` argument. Further this error message perhaps should read:
```
Argument 'types' must be an Iterable.
```
so as to explain what the requirement is and that it relates to the `types` argument.

Following this tests for all the `*Tuple*` constructors accepting an `Iterable` types can be written.

Comment (guilhermeleobas) [reply numba/types/containers.py]: For `Tuple` I had to put the check on `__new__` since there is a conditional there using `types`.
